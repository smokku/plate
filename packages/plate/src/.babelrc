{
  "presets": [
    "@babel/flow",
    ["@babel/env", {"modules": false}]
  ]
}
