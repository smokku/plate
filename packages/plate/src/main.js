// @flow
import {
  normalize,
  denormalize,
  type Schema as NormalizrSchema
} from 'normalizr'
import Immutable from 'seamless-immutable'
import type { AxiosPromise, AxiosXHRConfig } from 'axios'
import type { Store } from 'redux'
import _each from 'lodash/each'
import _camelCase from 'lodash/camelCase'
import _upperFirst from 'lodash/upperFirst'
import _mapValues from 'lodash/mapValues'
import {
  addEntities,
  getEntities,
  addResult,
  getResult,
  setStatus,
  getStatus
} from './redux'

import type { ReduxState } from './redux'

export { reducer, clearStore } from './redux'

export type MethodDescription = {
  url: string | ((...Array<any>) => string),
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE',
  data?: string | number | ((*) => *),
  headers?: { [string]: mixed },
  schema?: NormalizrSchema,
  selects?: (*) => *,
  returns?: (*) => *,
  preReq?: (req: AxiosXHRConfig<*>) => AxiosXHRConfig<*>
}

export type EntityName = string
export type EntityDescription = {
  [EntityName]: MethodDescription,
  schema?: NormalizrSchema
}

export type PlateStateSelector = (ReduxState, ...Array<mixed>) => boolean
export type PlateSelector = {
  (ReduxState, ...Array<mixed>): mixed,
  isProcessing: PlateStateSelector,
  isSuccess: PlateStateSelector,
  isError: PlateStateSelector
}
export const selectors: { [string]: PlateSelector } = {}
export type PlateAction = (...Array<mixed>) => Promise<mixed>
export const actions: { [string]: PlateAction } = {}

export type PlateSchema = {
  [entity: string]: {
    [endPoint: string]: MethodDescription,
    schema?: NormalizrSchema
  }
}

const noop = () => {}

export function configure (
  store: Store<*, *, *>,
  schema: PlateSchema,
  api: { [string]: (string, {}) => AxiosPromise<*> }
) {
  /*
   * This double for-each iterates over schema mapping
   * EntityDescription + MethodDescription to an action + selector
   * of the form `entityNameMethodName(args)` which are exported from this module.
   */
  _each(
    schema,
    (entityDescription: EntityDescription, entityName: EntityName) => {
      _each(
        entityDescription,
        (methodDescription: MethodDescription, methodName: string) => {
          /* skip common schema */
          if (methodName === 'schema') return

          const selectorPrefix = `${_camelCase(entityName)}${_upperFirst(
            _camelCase(methodName)
          )}`
          const methodSchema: NormalizrSchema =
            methodDescription.schema || entityDescription.schema || {}

          /* Maps `args` from MethodDescription */
          function mapArgsDescription (name, prop, args: Array<mixed>) {
            switch (typeof prop) {
              /* No parameter needed. */
              case 'undefined':
                break
              /* If `prop` is a function, just use it to map call arguments to actual value. */
              case 'function':
                return prop(...args)
              /* If `prop` is a string, this means prop is given as a property of
                  a first `args` argument, like: `entityMethod({prop: value})` */
              case 'string':
                if (args[0] && typeof args[0] === 'object') {
                  const [params] = args
                  if (Object.prototype.hasOwnProperty.call(params, prop)) {
                    return params[prop]
                  }
                  throw new TypeError(
                    `${entityName}/${methodName} 'args/${name}' requires '${prop}' property on ${selectorPrefix} argument.`
                  )
                } else {
                  throw new TypeError(
                    `${entityName}/${methodName} 'args/${name}' type 'string' requires ${selectorPrefix} ` +
                      `to be called with Object, but called with ${typeof args[0]}.`
                  )
                }
              /* If `data` is a number, this means data is given as an argument to call,
                  like: `entityMethod(entityId, requestData)` */
              case 'number':
                if (typeof args[prop] !== 'undefined') {
                  return args[prop]
                }
                throw new TypeError(
                  `${entityName}/${methodName} 'args/${name}' type 'number' requires ${selectorPrefix} ` +
                    `to be called with at least ${prop + 1} arguments.`
                )
              default:
                throw new TypeError(
                  `${entityName}/${methodName} 'args/${name}' type '${typeof prop}' not supported.`
                )
            }
            return undefined
          }

          /* Create a function used to do actual Axios request when needed */
          const fetchFunction = (...args) => {
            let {
              url,
              method = 'GET', // eslint-disable-line prefer-const
              headers = {}, // eslint-disable-line prefer-const
              data,
              preReq // eslint-disable-line prefer-const
            } = methodDescription

            /* If `url` is a function use it to map args to a real URL, if not, take url as is. */
            if (typeof url === 'function') {
              url = url(...args)
            }

            /* Axios request data */
            data = mapArgsDescription('data', data, args)

            store.dispatch(setStatus(selectorPrefix, args, 'PROCESSING'))

            /* Finally do an actual Axios call, mapping the response */
            const axiosMethod = method.toLowerCase()
            if (typeof api[axiosMethod] === 'function') {
              let axiosArgs: any = []
              switch (axiosMethod) {
                case 'post':
                case 'put':
                case 'patch':
                  axiosArgs = [data]
                  break
                default:
                // no data to post
              }
              let config = {}
              if (Object.keys(headers).length > 0) {
                config.headers = headers
              }
              if (typeof preReq === 'function') {
                config = preReq(config)
              }
              if (Object.keys(config).length > 0) {
                axiosArgs.push(config)
              }

              return api[axiosMethod](url, ...axiosArgs).then(
                ({ data: responseData }: { data: mixed }) => {
                  if (responseData && typeof responseData === 'object') {
                    const normalized = normalize(responseData, methodSchema)
                    store.dispatch(
                      addEntities(selectorPrefix, normalized.entities)
                    )
                    store.dispatch(
                      addResult(selectorPrefix, args, normalized.result)
                    )
                    store.dispatch(setStatus(selectorPrefix, args, 'SUCCESS'))
                    return normalized.result
                  }
                  // cannot denormalize non-object type - pass on
                  return responseData
                },
                error => {
                  store.dispatch(addResult(selectorPrefix, args, null))
                  store.dispatch(
                    setStatus(
                      selectorPrefix,
                      args,
                      'ERROR',
                      error.response?.data || error.message
                    )
                  )
                  return Promise.reject(error)
                }
              )
            }
            throw new TypeError(
              `${entityName}/${methodName} invalid method '${method}'.`
            )
          }

          /* Export generated fetchFunction(args) as explicit action
           * and wrapped in value selector calling fetchFunction as needed */
          actions[selectorPrefix] = fetchFunction
          selectors[selectorPrefix] = (state: ReduxState, ...args) => {
            const { selects: selector, returns: mangle } = methodDescription

            let result = getResult(state, selectorPrefix, args)

            if (
              typeof result === 'undefined' &&
              getStatus(state, selectorPrefix, args)?.status !== 'PROCESSING'
            ) {
              /* fetch in next tick to prevent state update loops */
              setTimeout(() => {
                if (
                  getStatus(store.getState(), selectorPrefix, args)?.status !==
                  'PROCESSING'
                ) {
                  /* selectors are to be silent */
                  fetchFunction(...args).then(noop, noop)
                }
              }, 0)
            }

            /* support selecting One from queryAll results */
            if (!result && selector) {
              result = selector(...args)
            }

            const entities = getEntities(state)
            const ret = denormalize(result, methodSchema, entities)

            return mangle ? mangle(ret) : ret
          }
          /* Additional selector to get the request status */
          const selectorPrefixStatus = `${selectorPrefix}_Status`
          selectors[selectorPrefixStatus] = (state: ReduxState, ...args) =>
            getStatus(state, selectorPrefix, args)
          selectors[selectorPrefixStatus].isProcessing = (
            state: ReduxState,
            ...args
          ) => getStatus(state, selectorPrefix, args)?.status === 'PROCESSING'
          selectors[selectorPrefixStatus].isSuccess = (
            state: ReduxState,
            ...args
          ) => getStatus(state, selectorPrefix, args)?.status === 'SUCCESS'
          selectors[selectorPrefixStatus].isError = (
            state: ReduxState,
            ...args
          ) => getStatus(state, selectorPrefix, args)?.status === 'ERROR'
        }
      )
    }
  )

  /* istanbul ignore next */
  if (process.env.NODE_ENV === 'development') {
    const bindState = f => (...args) => f(store.getState(), ...args)
    const bindObject = v => {
      if (typeof v === 'object') return _mapValues(v, bindObject)
      if (typeof v === 'function') {
        const bound = bindState(v)
        _each(_mapValues(v, bindObject), (s, k) => {
          bound[k] = s
        })
        return bound
      }
      return v
    }
    window.plate = { actions, selectors: _mapValues(selectors, bindObject) }
  }
}
