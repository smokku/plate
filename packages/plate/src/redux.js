// @flow
/* eslint-disable unicorn/new-for-builtins */
import Immutable, { type Immutable as ImmutableType } from 'seamless-immutable'
import type { EntityName } from './main'

export type ReduxEntities = ImmutableType<{[EntityName]: ReduxEntities}>
export type ReduxActionType = string
export type ReduxAction = {
  type: ReduxActionType,
  source: string,
  payload: any
}

export type RequestStatus = 'PROCESSING' | 'SUCCESS' | 'ERROR'
export type ReduxStatus = {
  status: RequestStatus,
  updatedAt: number,
  error?: string
}

export type ReduxInternalState = {
  entities: ReduxEntities,
  statuses: ImmutableType<{[string]: ReduxStatus}>,
  results: ImmutableType<{[string]: ImmutableType<{[string]: mixed}>}>
}
export type ReduxState = {
  plate: ReduxInternalState
}

export const CLEAR_STORE: ReduxActionType = '@@plate/clear-store'
export const clearStore = () => ({
  type: CLEAR_STORE,
  source: 'internal',
  payload: null
})

export const ADD_ENTITIES: ReduxActionType = '@@plate/add-entities'
export const addEntities = (source: string, entities: ReduxEntities) => ({
  type: ADD_ENTITIES,
  source,
  payload: entities
})
export const getEntities = (state: ReduxState): ReduxEntities =>
  state.plate.entities

const serialize = JSON.stringify
export const ADD_RESULT: ReduxActionType = '@@plate/add-result'
export const addResult = (source: string, args: Array<*>, result: mixed) => ({
  type: ADD_RESULT,
  source,
  payload: {
    args: serialize(args),
    result
  }
})
export const getResult = (
  state: ReduxState,
  source: string,
  args: Array<*>
) => {
  if (typeof state !== 'object' || !state.plate) {
    throw new TypeError(
      'getResult(state, source, args) requires ReduxState as first argument'
    )
  }
  // $FlowFixMe: Flow does not yet support method or property calls in optional chains.
  return state.plate.results[source]?.[serialize(args)]
}

export const SET_STATUS: ReduxActionType = '@@plate/set-status'
export const setStatus = (
  source: string,
  args: Array<mixed>,
  status: RequestStatus,
  error?: string
) => ({
  type: SET_STATUS,
  source,
  payload: {
    args: serialize(args),
    status,
    error
  }
})
export const getStatus = (
  state: ReduxState,
  source: string,
  args: Array<mixed>
): ?ReduxStatus => {
  if (typeof state !== 'object' || !state.plate) {
    throw new TypeError(
      'getStatus(state, source, args) requires ReduxState as first argument'
    )
  }
  // $FlowFixMe: Flow does not yet support method or property calls in optional chains.
  return state.plate.statuses[source]?.[serialize(args)]
}

export const initialState: ReduxInternalState = {
  entities: Immutable({}),
  statuses: Immutable({}),
  results: Immutable({})
}

export function reducer (
  state: ReduxInternalState = initialState,
  action: ReduxAction
): ReduxInternalState {
  switch (action.type) {
    case CLEAR_STORE:
      return initialState

    case ADD_ENTITIES:
      return {
        ...state,
        entities: state.entities.merge(action.payload, {
          deep: false,
          merger: (oldVal, newVal) => oldVal ? oldVal.merge(newVal, { deep: false }) : newVal
        })
      }

    case ADD_RESULT:
      return {
        ...state,
        results: state.results.merge(
          {
            [action.source]: {
              [action.payload.args]: action.payload.result
            }
          },
          { deep: true }
        )
      }

    case SET_STATUS:
      return {
        ...state,
        statuses: state.statuses.merge(
          {
            [action.source]: {
              [action.payload.args]: {
                status: action.payload.status,
                updatedAt: Date.now(),
                error: action.payload.error
              }
            }
          },
          { deep: true }
        )
      }

    default:
      return state
  }
}
