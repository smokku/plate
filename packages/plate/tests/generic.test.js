/* eslint-disable max-len, global-require, no-shadow, sonarjs/no-duplicate-string */
import {createStore, combineReducers} from 'redux'
import mockAxios from 'jest-mock-axios'
import {schema} from 'normalizr'
import {configure, selectors, actions, reducer, clearStore} from '../src/main'

const store = createStore(combineReducers({plate: reducer}))

const fur = new schema.Entity('fur')
const gerbil = new schema.Entity('gerbil', {fur})
configure(
  store,
  {
    gerbils: {
      GetAll: {
        url: '/gerbils',
        schema: {gerbils: [gerbil]},
      },
      GetArray: {
        url: '/gerbils',
        schema: {gerbils: [gerbil]},
        returns: ({gerbils} = {}) => gerbils,
      },
      getOne: {
        url: (id) => `/gerbils/${id}`,
        schema: gerbil,
        selects: (id) => id,
      },
      CreateOne: {
        url: '/gerbils_v2',
        method: 'POST',
        data: 'gerbilDescription',
        headers: {'X-Requested-With': 'Good app'},
      },
      setOne: {
        url: '/v7/gerbils',
        method: 'PUT',
        data: 3,
        schema: gerbil,
      },
      removeOne: {
        url: (id) => `/v7/gerbils/${id}`,
        method: 'DELETE',
        schema: gerbil,
      },
    },
    furs: {
      queryAll: {
        url: (q) => `/furs${q ? `?q=${q}` : ''}`,
        schema: {furs: [fur]},
      },
      getAuthed: {
        url: `/sum/ep`,
        preReq: (req) => ({
          ...req,
          headers: {Authorization: 'Bearer very-sikret-get-hasz'},
        }),
      },
      postAuthed: {
        url: `/sum/ep`,
        method: 'POST',
        preReq: (req) => ({
          ...req,
          headers: {Authorization: 'Bearer very-sikret-post-hasz'},
        }),
        data: () => 'sum data',
      },
      borked: {
        url: '???',
        // $FlowFixMe: testing invalid value
        data: false,
      },
      missing: {
        url: '---',
        // $FlowFixMe: testing invalid value
        method: 'run-forest-run',
      },
      schema: {},
    },
  },
  mockAxios
)

const mockGerbilsAllGet = {
  gerbils: [
    {id: '1', name: 'Jerry', fur: {id: '1', type: 'Fluffy'}},
    {id: '2', name: 'Terry', fur: {id: '1', color: 'White'}},
    {id: '3', name: 'Marry', fur: {id: '2', color: 'Orange'}},
    {id: '4', name: 'Jeshua'},
  ],
}

const mockGerbilOneGet1 = mockGerbilsAllGet.gerbils[0]
const mockGerbilOneGet2 = mockGerbilsAllGet.gerbils[1]

describe('Plate', () => {
  beforeEach(() => {
    jest.useFakeTimers()
    store.dispatch(clearStore())
    mockAxios.reset()
  })

  it('fetches all', () => {
    let gerbils = selectors.gerbilsGetAll(store.getState())
    expect(gerbils).toBeUndefined()
    expect(setTimeout).toHaveBeenCalledTimes(1)
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 0)
    expect(mockAxios.get).toHaveBeenCalledTimes(0)
    jest.runAllTimers()
    expect(mockAxios.get).toHaveBeenCalledTimes(1)
    expect(mockAxios.get).toHaveBeenCalledWith('/gerbils')
    mockAxios.mockResponse({data: mockGerbilsAllGet})
    gerbils = selectors.gerbilsGetAll(store.getState())
    expect(mockAxios.get).toHaveBeenCalledTimes(1)
    expect(gerbils).toMatchSnapshot()
  })

  it('mangles return', () => {
    selectors.gerbilsGetArray(store.getState())
    jest.runAllTimers()
    expect(mockAxios.get).toHaveBeenCalledTimes(1)
    expect(mockAxios.get).toHaveBeenCalledWith('/gerbils')
    mockAxios.mockResponse({data: mockGerbilsAllGet})
    const gerbils = selectors.gerbilsGetArray(store.getState())
    expect(mockAxios.get).toHaveBeenCalledTimes(1)
    expect(gerbils).toMatchSnapshot()
  })

  it('fetches one', () => {
    let gerbil = selectors.gerbilsGetOne(store.getState(), '1')
    jest.runAllTimers()
    expect(gerbil).toBeUndefined()
    expect(mockAxios.get).toHaveBeenCalledTimes(1)
    expect(mockAxios.get).toHaveBeenCalledWith('/gerbils/1')
    mockAxios.mockResponse({data: mockGerbilOneGet1})
    gerbil = selectors.gerbilsGetOne(store.getState(), '1')
    jest.runAllTimers()
    expect(mockAxios.get).toHaveBeenCalledTimes(1)
    expect(gerbil).toEqual(mockGerbilOneGet1)
    gerbil = selectors.gerbilsGetOne(store.getState(), '2')
    jest.runAllTimers()
    expect(gerbil).toBeUndefined()
    expect(mockAxios.get).toHaveBeenCalledTimes(2)
    expect(mockAxios.get).toHaveBeenCalledWith('/gerbils/2')
    mockAxios.mockResponse({data: mockGerbilOneGet2})
    gerbil = selectors.gerbilsGetOne(store.getState(), '2')
    jest.runAllTimers()
    expect(mockAxios.get).toHaveBeenCalledTimes(2)
    expect(gerbil).toMatchSnapshot()
  })

  it('gets one from all', () => {
    const gerbils = selectors.gerbilsGetAll(store.getState())
    jest.runAllTimers()
    expect(gerbils).toBeUndefined()
    expect(mockAxios.get).toHaveBeenCalledTimes(1)
    expect(mockAxios.get).toHaveBeenCalledWith('/gerbils')
    mockAxios.mockResponse({data: mockGerbilsAllGet})
    const gerbil = selectors.gerbilsGetOne(store.getState(), '2')
    jest.runAllTimers()
    expect(mockAxios.get).toHaveBeenCalledTimes(2)
    expect(gerbil).toMatchSnapshot()
  })

  it('queries all', () => {
    let furs = selectors.fursQueryAll(store.getState(), 'preeeeetyyy')
    furs = selectors.fursQueryAll(store.getState(), 'preeeeetyyy')
    furs = selectors.fursQueryAll(store.getState(), 'preeeeetyyy')
    furs = selectors.fursQueryAll(store.getState(), 'preeeeetyyy')
    jest.runAllTimers()
    expect(furs).toBeUndefined()
    expect(mockAxios.get).toHaveBeenCalledTimes(1)
    expect(mockAxios.get).toHaveBeenCalledWith('/furs?q=preeeeetyyy')
    actions.fursQueryAll('lame')
    expect(mockAxios.get).toHaveBeenCalledTimes(2)
    expect(mockAxios.get).toHaveBeenCalledWith('/furs?q=lame')
  })

  it('queries many', () => {
    let furs1 = selectors.fursQueryAll(store.getState(), 'q1')
    const req1 = mockAxios.lastReqGet()
    let furs2 = selectors.fursQueryAll(store.getState(), 'q2')
    const req2 = mockAxios.lastReqGet()
    jest.runAllTimers()
    expect(furs1).toBeUndefined()
    expect(mockAxios.get).toHaveBeenCalledTimes(2)
    expect(mockAxios.get).toHaveBeenCalledWith('/furs?q=q1')
    expect(mockAxios.get).toHaveBeenCalledWith('/furs?q=q2')

    const furs = mockGerbilsAllGet.gerbils
      .map((gerbil) => gerbil.fur)
      .filter((fur) => fur)
    mockAxios.mockResponse({data: {furs}}, req1)
    furs1 = selectors.fursQueryAll(store.getState(), 'q1')
    furs2 = selectors.fursQueryAll(store.getState(), 'q2')
    jest.runAllTimers()
    expect(mockAxios.get).toHaveBeenCalledTimes(2)
    expect(furs1).toMatchSnapshot()
    expect(furs2).toBeUndefined()

    mockAxios.mockResponse({data: {furs: [furs[furs.length - 1]]}}, req2)
    const furs12 = selectors.fursQueryAll(store.getState(), 'q1')
    furs2 = selectors.fursQueryAll(store.getState(), 'q2')
    jest.runAllTimers()
    expect(mockAxios.get).toHaveBeenCalledTimes(2)
    expect(furs12).toEqual(furs1)
    expect(furs2).toMatchSnapshot()

    const furs13 = selectors.fursQueryAll(store.getState(), 'q1')
    const furs23 = selectors.fursQueryAll(store.getState(), 'q2')
    jest.runAllTimers()
    expect(mockAxios.get).toHaveBeenCalledTimes(2)
    expect(furs13).toEqual(furs1)
    expect(furs23).toEqual(furs2)
  })

  it('creates one', () => {
    expect.assertions(3)
    const create = actions.gerbilsCreateOne({gerbilDescription: 'funny one'})
    expect(mockAxios.post).toHaveBeenCalledTimes(1)
    expect(mockAxios.post).toHaveBeenCalledWith('/gerbils_v2', 'funny one', {
      headers: {'X-Requested-With': 'Good app'},
    })
    mockAxios.mockResponse('OK')
    expect(create).resolves.toEqual({})
  })

  it('sets one', () => {
    expect.assertions(3)
    const create = actions.gerbilsSetOne('one', 'dos', 'trzy', 'blorg')
    expect(mockAxios.put).toHaveBeenCalledTimes(1)
    expect(mockAxios.put).toHaveBeenCalledWith('/v7/gerbils', 'blorg')
    mockAxios.mockResponse({data: mockGerbilOneGet1})
    expect(create).resolves.toEqual('1')
  })

  it('removes one', () => {
    expect.assertions(3)
    const create = actions.gerbilsRemoveOne('one')
    expect(mockAxios.delete).toHaveBeenCalledTimes(1)
    expect(mockAxios.delete).toHaveBeenCalledWith('/v7/gerbils/one')
    mockAxios.mockResponse({data: mockGerbilOneGet1})
    expect(create).resolves.toEqual('1')
  })

  it('manages statuses', () => {
    let status = selectors.gerbilsGetAll_Status(store.getState())
    let statusIsProcessing = selectors.gerbilsGetAll_Status.isProcessing(
      store.getState()
    )
    let statusIsSuccess = selectors.gerbilsGetAll_Status.isSuccess(
      store.getState()
    )
    let statusIsError = selectors.gerbilsGetAll_Status.isError(store.getState())
    expect(status).toBeUndefined()
    expect(statusIsProcessing).toEqual(false)
    expect(statusIsSuccess).toEqual(false)
    expect(statusIsError).toEqual(false)
    actions.gerbilsGetAll()
    status = selectors.gerbilsGetAll_Status(store.getState())
    statusIsProcessing = selectors.gerbilsGetAll_Status.isProcessing(
      store.getState()
    )
    statusIsSuccess = selectors.gerbilsGetAll_Status.isSuccess(store.getState())
    statusIsError = selectors.gerbilsGetAll_Status.isError(store.getState())
    expect(typeof status).toEqual('object')
    expect(status?.status).toBe('PROCESSING')
    expect(typeof status?.updatedAt).toEqual('number')
    expect(statusIsProcessing).toEqual(true)
    expect(statusIsSuccess).toEqual(false)
    expect(statusIsError).toEqual(false)
    mockAxios.mockResponse({data: mockGerbilsAllGet})
    status = selectors.gerbilsGetAll_Status(store.getState())
    statusIsProcessing = selectors.gerbilsGetAll_Status.isProcessing(
      store.getState()
    )
    statusIsSuccess = selectors.gerbilsGetAll_Status.isSuccess(store.getState())
    statusIsError = selectors.gerbilsGetAll_Status.isError(store.getState())
    expect(status?.status).toBe('SUCCESS')
    expect(statusIsProcessing).toEqual(false)
    expect(statusIsSuccess).toEqual(true)
    expect(statusIsError).toEqual(false)
  })

  it('fails selector parameters', () => {
    expect(() => selectors.gerbilsGetAll()).toThrow(
      /getResult\(state, source, args\) requires ReduxState as first argument/
    )
    expect(() => selectors.gerbilsGetAll_Status()).toThrow(
      /getStatus\(state, source, args\) requires ReduxState as first argument/
    )
  })

  it('fails action parameters', () => {
    expect(() => actions.gerbilsCreateOne()).toThrow(
      /gerbils\/CreateOne 'args\/data' type 'string' requires gerbilsCreateOne to be called with Object, but called with undefined/
    )
    expect(() => actions.gerbilsCreateOne({gerbils: ['One', 'Two']})).toThrow(
      /gerbils\/CreateOne 'args\/data' requires 'gerbilDescription' property on gerbilsCreateOne argument/
    )
    expect(() => actions.gerbilsSetOne()).toThrow(
      /gerbils\/setOne 'args\/data' type 'number' requires gerbilsSetOne to be called with at least 4 arguments/
    )
    expect(() => actions.fursBorked()).toThrow(
      /furs\/borked 'args\/data' type 'boolean' not supported/
    )
    expect(() => actions.fursMissing()).toThrow(
      /furs\/missing invalid method 'run-forest-run'/
    )
    jest.runAllTimers()
    expect(mockAxios.get).not.toHaveBeenCalled()
    expect(mockAxios.post).not.toHaveBeenCalled()
    expect(mockAxios.put).not.toHaveBeenCalled()
    expect(mockAxios.delete).not.toHaveBeenCalled()
  })

  it('fails request', () => {
    let gerbils = selectors.gerbilsGetAll(store.getState())
    expect(gerbils).toBeUndefined()
    jest.runAllTimers()
    mockAxios.mockError({message: 'Łolaboga!'})
    gerbils = selectors.gerbilsGetAll(store.getState())
    const status = selectors.gerbilsGetAll_Status(store.getState())
    const error = selectors.gerbilsGetAll_Status.isError(store.getState())
    jest.runAllTimers()
    expect(mockAxios.get).toHaveBeenCalledTimes(1)
    expect(gerbils).toEqual({})
    expect(status).toHaveProperty('status', 'ERROR')
    expect(status).toHaveProperty('error', 'Łolaboga!')
    expect(error).toEqual(true)
  })

  it('passes non-object response', () => {
    const gerbils = actions.gerbilsGetAll()
    mockAxios.mockResponse({data: 'I can has string reply'})
    expect(gerbils).resolves.toEqual('I can has string reply')
  })

  it('handles preReq functions', () => {
    actions.fursGetAuthed()
    expect(mockAxios.get).toHaveBeenCalledWith('/sum/ep', {
      headers: {Authorization: 'Bearer very-sikret-get-hasz'},
    })
    expect(mockAxios.post).toHaveBeenCalledTimes(0)
    actions.fursPostAuthed()
    expect(mockAxios.post).toHaveBeenCalledWith('/sum/ep', 'sum data', {
      headers: {Authorization: 'Bearer very-sikret-post-hasz'},
    })
    expect(mockAxios.get).toHaveBeenCalledTimes(1)
  })
})
