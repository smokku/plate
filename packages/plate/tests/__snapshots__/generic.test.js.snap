// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plate fetches all 1`] = `
Object {
  "gerbils": Array [
    Object {
      "fur": Object {
        "color": "White",
        "id": "1",
        "type": "Fluffy",
      },
      "id": "1",
      "name": "Jerry",
    },
    Object {
      "fur": Object {
        "color": "White",
        "id": "1",
        "type": "Fluffy",
      },
      "id": "2",
      "name": "Terry",
    },
    Object {
      "fur": Object {
        "color": "Orange",
        "id": "2",
      },
      "id": "3",
      "name": "Marry",
    },
    Object {
      "id": "4",
      "name": "Jeshua",
    },
  ],
}
`;

exports[`Plate fetches one 1`] = `
Object {
  "fur": Object {
    "color": "White",
    "id": "1",
  },
  "id": "2",
  "name": "Terry",
}
`;

exports[`Plate gets one from all 1`] = `
Object {
  "fur": Object {
    "color": "White",
    "id": "1",
    "type": "Fluffy",
  },
  "id": "2",
  "name": "Terry",
}
`;

exports[`Plate mangles return 1`] = `
Array [
  Object {
    "fur": Object {
      "color": "White",
      "id": "1",
      "type": "Fluffy",
    },
    "id": "1",
    "name": "Jerry",
  },
  Object {
    "fur": Object {
      "color": "White",
      "id": "1",
      "type": "Fluffy",
    },
    "id": "2",
    "name": "Terry",
  },
  Object {
    "fur": Object {
      "color": "Orange",
      "id": "2",
    },
    "id": "3",
    "name": "Marry",
  },
  Object {
    "id": "4",
    "name": "Jeshua",
  },
]
`;

exports[`Plate queries many 1`] = `
Object {
  "furs": Array [
    Object {
      "color": "White",
      "id": "1",
      "type": "Fluffy",
    },
    Object {
      "color": "White",
      "id": "1",
      "type": "Fluffy",
    },
    Object {
      "color": "Orange",
      "id": "2",
    },
  ],
}
`;

exports[`Plate queries many 2`] = `
Object {
  "furs": Array [
    Object {
      "color": "Orange",
      "id": "2",
    },
  ],
}
`;
